Include/internal/pycore_stackref.h

```diff
@@ -0,0 +1,189 @@
+#ifndef Py_INTERNAL_STACKREF_H
+#define Py_INTERNAL_STACKREF_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include <stddef.h>
+#include "pycore_gc.h"
+
+// Mark an object as supporting deferred reference counting. This is a no-op
+// in the default (with GIL) build. Objects that use deferred reference
+// counting should be tracked by the GC so that they are eventually collected.
+extern void _PyObject_SetDeferredRefcount(PyObject *op);
+
+static inline int
+_PyObject_HasDeferredRefcount(PyObject *op)
+{
+#ifdef Py_GIL_DISABLED
+    return (op->ob_gc_bits & _PyGC_BITS_DEFERRED) != 0;
+#else
+    return 0;
+#endif
+}
+
+typedef union {
+    uintptr_t bits;
+} _PyStackRef;
+
+static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 };
+
+#ifdef Py_GIL_DISABLED
+    #define Py_TAG_DEFERRED (1)
+    #define Py_TAG (Py_TAG_DEFERRED)
+#else
+    #define Py_TAG 0
+#endif
+
+// Gets a PyObject * from a _PyStackRef
+#if defined(Py_GIL_DISABLED)
+    static inline PyObject *
+    PyStackRef_Get(_PyStackRef tagged) {
+        PyObject *cleared = ((PyObject *)((tagged).bits & (~Py_TAG)));
+        return cleared;
+    }
+#else
+    #define PyStackRef_Get(tagged) ((PyObject *)(uintptr_t)((tagged).bits))
+#endif
```

Comment:
Let's clean-up and format these macros and functions in more common CPython styles.

* Braces on their own line
* We don't generally indent functions inside `ifdef`s
* The below is the style I've seen more commonly for nested `#define`s

```suggestion
#if defined(Py_GIL_DISABLED)
static inline PyObject *
PyStackRef_Get(_PyStackRef tagged)
{
    return (PyObject *)(tagged.bits & ~Py_TAG_DEFERRED);
}
#else
#  define PyStackRef_Get(tagged) ((PyObject *)((tagged).bits))
#endif
```